sycl/include/syclcompat/math.hpp

```diff
@@ -109,12 +109,33 @@ clamp(sycl::marray<T, Size> val, sycl::marray<T, Size> min_val,
 template <typename VecT, class BinaryOperation, class = void>
 class vectorized_binary {
 public:
-  inline VecT operator()(VecT a, VecT b, const BinaryOperation binary_op) {
+  inline unsigned operator()(VecT a, VecT b, const BinaryOperation binary_op) {
     VecT v4;
     for (size_t i = 0; i < v4.size(); ++i) {
       v4[i] = binary_op(a[i], b[i]);
     }
-    return v4;
+    return sycl::vec<unsigned, 1>(v4.template as<sycl::vec<unsigned, 1>>())[0];
+  }
+};
+
+// Vectorized_binary for logical operations
+template <typename VecT, class BinaryOperation>
+class vectorized_binary<
+    VecT, BinaryOperation,
+    std::enable_if_t<std::is_same_v<
+        bool, decltype(std::declval<BinaryOperation>()(
+                  std::declval<typename VecT::element_type>(),
+                  std::declval<typename VecT::element_type>()))>>> {
+public:
+  inline unsigned operator()(VecT a, VecT b, const BinaryOperation binary_op) {
+    unsigned result = 0;
+    constexpr size_t elem_size = sizeof(typename VecT::element_type);
```

Comment:
The boolean logic below works only for elem_size 1,2,3 ... 5,6,7, and so on.

`if elem_size == 4`, `bool_mask` is `0`, so the result would always be false.

Since we are almost certain that this operation should only be used with 1 and 2-byte types, I'd recommend to document it and guard it.

Alternatively, you could change the mask to something like `0xFFFFFFFF >> ((4 - (elem_size % 4)) * 8)` but it's rather difficult to read, just  guard against cases that should not be reached.

Reply:
Done I've added a static assert, thanks for the hint @Alcpz.